mqt/ddsim/bindings.cpp

```diff
@@ -141,45 +141,56 @@ void dumpTensorNetwork(const py::object& circ, const std::string& filename) {
     qc->dump(ofs, qc::Format::Tensor);
 }
 
+template<class Simulator>
+void addCommonSimulatorMethods(py::class_<Simulator>& cls) {
```

Comment:
Would it be cleaner (and possible) to have like a templated factory for the simulator bindings?

The following code is just to illustrate the idea, I don't think it will compile :S

```cpp
template<class Sim>
auto createSimulator(pymodule m, const std::string name)  {
    auto sim = py::class_<Sim>(m, name);
   sim.def("get_number_of_qubits", &Simulator::getNumberOfQubits, "Get the number of qubits")
            .def("get_name", &Simulator::getName, "Get the name of the simulator")
            .def("statistics", &Simulator::additionalStatistics, "Get additional statistics provided by the simulator")
            .def("get_active_vector_node_count", &Simulator::getActiveNodeCount, "Get the number of active vector nodes, i.e., the number of vector DD nodes in the unique table with a non-zero reference count.")
            .def("get_active_matrix_node_count", &Simulator::getMatrixActiveNodeCount, "Get the number of active matrix nodes, i.e., the number of matrix DD nodes in the unique table with a non-zero reference count.")
            .def("get_max_vector_node_count", &Simulator::getMaxNodeCount, "Get the maximum number of (active) vector nodes, i.e., the maximum number of vector DD nodes in the unique table at any point during the simulation.")
            .def("get_max_matrix_node_count", &Simulator::getMaxMatrixNodeCount, "Get the maximum number of (active) matrix nodes, i.e., the maximum number of matrix DD nodes in the unique table at any point during the simulation.")
            .def("set_tolerance", &Simulator::setTolerance, "tol"_a, "Set the tolerance for the DD package.");

    if constexpr (std::is_same_v<Sim, UnitarySimulator<>>) {
        sim.def("construct", &Simulator::construct, "Construct the DD representing the unitary matrix of the circuit.");
    } else {
        sim.def("simulate", &Simulator::simulate, "shots"_a, "Simulate the circuit and return the result as a dictionary of counts.");
        sim.def("get_vector", &Simulator::getVectorComplex, "Get the state vector resulting from the simulation.");
    }
    return sim;
}
```

And used like

```c++
auto circuitSimulator = createSimulator<CircuitSimulator>(m, "CircuitSimulator");
circuitSimulator.def(py::init<>(&createSimulator<CircuitSimulator<>>),
                         "circ"_a,
                         "approximation_step_fidelity"_a = 1.,
                         "approximation_steps"_a         = 1,
                         "approximation_strategy"_a      = "fidelity",
                         "seed"_a                        = -1);
```

This seems cleaner to me, but I'm not sure if it works as intended (especially returning the objects seems like it could break things). What do you think?

Reply:
Done in https://github.com/cda-tum/ddsim/pull/202/commits/74aefdb3d34bf31e76f1014077d54663ddb4d352
Thanks for the suggestion. For the moment, it saves only a couple of lines. In the long run that seems like a nice option.
If you are satisfied with the result, I'd probably favor keeping this change.

Reply:
That looks very good :)